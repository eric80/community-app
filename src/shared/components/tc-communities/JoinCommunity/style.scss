@import "~styles/mixins";

.joining:hover {
  background: $tc-white;
  cursor: default;
}

.joining div p {
  display: inline-block;
  margin-right: 12px;
  vertical-align: top;
}

.loadingIndicator {
  display: inline-block;
  height: 32px;
  padding-top: 4px;
  width: 32px;
}

.modalTitle {
  color: $tc-black;
  text-align: center;
}

.modalMsg {
  @include tc-body-md;

  color: #787d81;
  margin: 24px 0;
  text-align: center;
}

.placeholder {
  height: 24px;
}
