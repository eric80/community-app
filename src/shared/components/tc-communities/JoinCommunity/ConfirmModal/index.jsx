/**
 * The modal asking visitor to confirm that he wants to join community where
 * JoinCommunity component is rendered. If visitor is authenticated it shows
 * Join / Cancel buttons that do what they are supposed to do; otherwise it
 * shows Register & Join / Log In & Join / Cancel buttons, first of which get
 * visitor to registration / login flows, with return url making the join to
 * happen automatically.
 */

/* global window */

import config from 'utils/config';
import Modal from 'components/Modal';
import PT from 'prop-types';
import React from 'react';
import { PrimaryButton, SecondaryButton } from 'components/buttons';
import style from './style.scss';

export default function ConfirmModal({
  communityId,
  communityName,
  groupIds,
  join,
  resetJoinButton,
  token,
  userId,
}) {
  return (
    <Modal onCancel={resetJoinButton}>
      <div styleName="style.confirmMsg">
        { userId ? (
          <p>Do you want to join {communityName}?</p>
        ) : (
          <div>
            <p>You must be a Topcoder member before you can join the {communityName}.</p>
            <p>To join, login if you are already a member. If not, register first.</p>
          </div>
        )}
      </div>
      { userId ? (
        <div className={style.joinButtons}>
          <PrimaryButton
            onClick={() => join(token, groupIds[0], userId)}
          >Join</PrimaryButton>
          <SecondaryButton
            onClick={resetJoinButton}
          >Cancel</SecondaryButton>
        </div>
      ) : (
        <div className={style.loginButtons}>
          <PrimaryButton
            onClick={() => {
              const url = encodeURIComponent(
                `${window.location.href}?join=${groupIds[0]}`,
              );
              window.location = `${config.URL.AUTH}/member?retUrl=${url}&utm_source=${communityId}`;
            }}
          >Login</PrimaryButton>
          <PrimaryButton
            onClick={() => {
              let url = encodeURIComponent(
                `${window.location.href}?join=${groupIds[0]}`,
              );
              url = encodeURIComponent(
                `${config.URL.AUTH}/member?retUrl=${url}&utm_source=${communityId}`,
              );
              url = encodeURIComponent(url);
              window.location = `${config.URL.AUTH}/member/registration?retUrl=${url}&utm_source=${communityId}`;
            }}
          >Register</PrimaryButton>
          <SecondaryButton
            onClick={resetJoinButton}
          >Cancel</SecondaryButton>
        </div>
      )}
    </Modal>
  );
}

ConfirmModal.defaultProps = {
  token: null,
  userId: null,
};

ConfirmModal.propTypes = {
  communityId: PT.string.isRequired,
  communityName: PT.string.isRequired,
  groupIds: PT.arrayOf(PT.string).isRequired,
  join: PT.func.isRequired,
  resetJoinButton: PT.func.isRequired,
  token: PT.string,
  userId: PT.string,
};
