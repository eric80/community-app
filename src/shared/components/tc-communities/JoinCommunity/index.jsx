/**
 * Join Community component. It includes 'Join Community' button, which is
 * automatically hidden, when a user is already member of community. Button
 * text changes to the loading symbol, when the joining is underway. And a
 * modal is shown on success.
 */

import _ from 'lodash';
import LoadingIndicator from 'components/LoadingIndicator';
import Modal from 'components/Modal';
import PT from 'prop-types';
import React from 'react';
import { Button, PrimaryButton } from 'topcoder-react-ui-kit';
import { COMPOSE } from 'react-css-super-themr';
import style from './style.scss';

import ConfirmModal from './ConfirmModal';

export const STATE = {
  CONFIRM_JOIN: 'confirm-join',
  DEFAULT: 'default',
  HIDDEN: 'hidden',
  JOINED: 'joined',
  JOINING: 'joining',
};

export default function JoinCommunity({
  communityId,
  communityName,
  groupIds,
  hideJoinButton,
  join,
  label,
  resetJoinButton,
  showJoinConfirmModal,
  state,
  theme,
  token,
  userId,
}) {
  if (state === STATE.HIDDEN) return <div styleName="style.placeholder" />;
  return (
    <div
      className={theme.container}
    >
      <Button
        onClick={() => {
          switch (state) {
            case STATE.JOINED:
            case STATE.JOINING:
              return;
            default:
          }
          showJoinConfirmModal();
        }}
        className={state === STATE.JOINING ? style.joining : ''}
        {...(theme.link ? { theme: theme.link, composeContextTheme: COMPOSE.SWAP } : {})}
      >
        { state === STATE.JOINING ? (
          <div className={style.joiningContainer}>
            <span>Joining...</span>
            <LoadingIndicator theme={{ container: style.loadingIndicator }} />
          </div>
        ) : label}
      </Button>
      { state === STATE.JOINED ? (
        <Modal onCancel={hideJoinButton}>
          <h1 styleName="style.modalTitle">Congratulations!</h1>
          <p styleName="style.modalMsg">You have joined the {communityName}!</p>
          <PrimaryButton
            onClick={hideJoinButton}
            theme={{
              button: style.returnToCommunityButton,
            }}
          >Return to the Community</PrimaryButton>
        </Modal>
      ) : null}
      { state === STATE.CONFIRM_JOIN ? (
        <ConfirmModal
          communityId={communityId}
          communityName={communityName}
          groupIds={groupIds}
          join={join}
          resetJoinButton={resetJoinButton}
          token={token}
          userId={userId}
        />
      ) : null}
    </div>
  );
}

JoinCommunity.defaultProps = {
  groupIds: [''],
  label: 'Join Community',
  theme: {},
  token: null,
  userId: null,
};

JoinCommunity.propTypes = {
  communityId: PT.string.isRequired,
  communityName: PT.string.isRequired,
  groupIds: PT.arrayOf(PT.string),
  hideJoinButton: PT.func.isRequired,
  join: PT.func.isRequired,
  label: PT.string,
  resetJoinButton: PT.func.isRequired,
  showJoinConfirmModal: PT.func.isRequired,
  state: PT.oneOf(_.values(STATE)).isRequired,
  theme: PT.shape(),
  token: PT.string,
  userId: PT.string,
};
