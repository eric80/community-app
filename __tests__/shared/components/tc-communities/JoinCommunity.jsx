import JoinCommunity from 'components/tc-communities/JoinCommunity';
import React from 'react';
import TU from 'react-dom/test-utils';
import {
  renderDom,
  shallowSnapshot,
  simulate,
} from 'topcoder-react-utils/jest-utils';

const hideJoinButton = jest.fn();
const join = jest.fn();
const resetJoinButton = jest.fn();
const showJoinConfirmModal = jest.fn();

const mockDatas = [
  {
    communityName: 'ios',
    groupId: '1',
    hideJoinButton,
    join,
    resetJoinButton,
    showJoinConfirmModal,
    state: 'confirm-join',
    token: 'token',
    userId: 'useId',
  }, {
    communityName: 'ios',
    groupId: '1',
    hideJoinButton,
    join,
    resetJoinButton,
    showJoinConfirmModal,
    state: 'joining',
    token: 'token',
    userId: 'useId',
  }, {
    communityName: 'ios',
    groupId: '1',
    hideJoinButton,
    join,
    resetJoinButton,
    showJoinConfirmModal,
    state: 'hidden',
    token: 'token',
    userId: 'useId',
  }, {
    communityName: 'ios',
    groupId: '1',
    hideJoinButton,
    join,
    resetJoinButton,
    showJoinConfirmModal,
    state: 'joined',
    token: 'token',
    userId: 'useId',
  }, {
    communityName: 'ios',
    groupId: '1',
    hideJoinButton,
    join,
    resetJoinButton,
    showJoinConfirmModal,
    state: 'confirm-join',
    userId: 'useId',
  },
];

test('Matches shallow shapshot', () => {
  mockDatas.forEach(data => shallowSnapshot(<JoinCommunity {...data} />));
});

test('click confirm-join', () => {
  const dom = renderDom(<JoinCommunity {...mockDatas[0]} />);
  const matches = TU.scryRenderedDOMComponentsWithTag(dom, 'button');
  matches.forEach(m => simulate.click(m));
  expect(join).toHaveBeenCalled();
});

test('click joined', () => {
  const instance = renderDom(<JoinCommunity {...mockDatas[3]} />);
  const matches = TU.scryRenderedDOMComponentsWithTag(instance, 'button');
  TU.Simulate.click(matches[0]);
});

test('click without token', () => {
  const instance = renderDom(<JoinCommunity {...mockDatas[4]} />);
  const matches = TU.scryRenderedDOMComponentsWithTag(instance, 'button');
  TU.Simulate.click(matches[0]);
});
