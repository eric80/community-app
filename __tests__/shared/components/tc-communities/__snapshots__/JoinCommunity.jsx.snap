// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<div
  className={undefined}
>
  <ThemedButton
    active={false}
    className=""
    composeAdhocTheme="deeply"
    composeContextTheme="softly"
    disabled={false}
    enforceA={false}
    mapThemrProps={[Function]}
    onClick={[Function]}
    onMouseDown={null}
    openNewTab={false}
    replace={false}
    size={null}
    themePriority="adhoc-context-default"
    to={null}
  >
    Join Community
  </ThemedButton>
  <ConfirmModal
    communityId={undefined}
    communityName="ios"
    groupIds={
      Array [
        "",
      ]
    }
    join={[Function]}
    resetJoinButton={[Function]}
    token="token"
    userId="useId"
  />
</div>
`;

exports[`Matches shallow shapshot 2`] = `
<div
  className={undefined}
>
  <ThemedButton
    active={false}
    className="src-shared-components-tc-communities-JoinCommunity-style___joining___rq-m2x"
    composeAdhocTheme="deeply"
    composeContextTheme="softly"
    disabled={false}
    enforceA={false}
    mapThemrProps={[Function]}
    onClick={[Function]}
    onMouseDown={null}
    openNewTab={false}
    replace={false}
    size={null}
    themePriority="adhoc-context-default"
    to={null}
  >
    <div
      className="src-shared-components-tc-communities-JoinCommunity-style___joiningContainer___2P4Cvw"
    >
      <span>
        Joining...
      </span>
      <ThemedLoadingIndicator
        composeAdhocTheme="deeply"
        composeContextTheme="softly"
        mapThemrProps={[Function]}
        theme={
          Object {
            "container": "src-shared-components-tc-communities-JoinCommunity-style___loadingIndicator___2RP1uY",
          }
        }
        themePriority="adhoc-context-default"
      />
    </div>
  </ThemedButton>
</div>
`;

exports[`Matches shallow shapshot 3`] = `
<div
  className="src-shared-components-tc-communities-JoinCommunity-___style__placeholder___1XfW2"
/>
`;

exports[`Matches shallow shapshot 4`] = `
<div
  className={undefined}
>
  <ThemedButton
    active={false}
    className=""
    composeAdhocTheme="deeply"
    composeContextTheme="softly"
    disabled={false}
    enforceA={false}
    mapThemrProps={[Function]}
    onClick={[Function]}
    onMouseDown={null}
    openNewTab={false}
    replace={false}
    size={null}
    themePriority="adhoc-context-default"
    to={null}
  >
    Join Community
  </ThemedButton>
  <ThemedModal
    composeAdhocTheme="deeply"
    composeContextTheme="softly"
    mapThemrProps={[Function]}
    onCancel={[Function]}
    theme={Object {}}
    themePriority="adhoc-context-default"
  >
    <h1
      className="src-shared-components-tc-communities-JoinCommunity-___style__modalTitle___1U_iw"
    >
      Congratulations!
    </h1>
    <p
      className="src-shared-components-tc-communities-JoinCommunity-___style__modalMsg___25f_a"
    >
      You have joined the 
      ios
      !
    </p>
    <ThemedButton
      active={false}
      composeAdhocTheme="deeply"
      composeContextTheme="softly"
      disabled={false}
      enforceA={false}
      mapThemrProps={[Function]}
      onClick={[Function]}
      onMouseDown={null}
      openNewTab={false}
      replace={false}
      size={null}
      theme={
        Object {
          "button": "src-shared-components-tc-communities-JoinCommunity-style___returnToCommunityButton___1YlEiP",
        }
      }
      themePriority="adhoc-context-default"
      to={null}
    >
      Return to the Community
    </ThemedButton>
  </ThemedModal>
</div>
`;

exports[`Matches shallow shapshot 5`] = `
<div
  className={undefined}
>
  <ThemedButton
    active={false}
    className=""
    composeAdhocTheme="deeply"
    composeContextTheme="softly"
    disabled={false}
    enforceA={false}
    mapThemrProps={[Function]}
    onClick={[Function]}
    onMouseDown={null}
    openNewTab={false}
    replace={false}
    size={null}
    themePriority="adhoc-context-default"
    to={null}
  >
    Join Community
  </ThemedButton>
  <ConfirmModal
    communityId={undefined}
    communityName="ios"
    groupIds={
      Array [
        "",
      ]
    }
    join={[Function]}
    resetJoinButton={[Function]}
    token={null}
    userId="useId"
  />
</div>
`;
